File: go/tree/107_level_order_bottom.go
package tree

// levelOrderBottom 从低向上层次遍历
// 层序遍历后，反转
func levelOrderBottom(root *TreeNode) [][]int {

	if root == nil {
		return nil
	}

	ans := make([][]int, 0)

	queue := make([]*TreeNode, 0)
	queue = append(queue, root)

	for len(queue) > 0 {
		level := make([]int, 0)

		// 遍历当前这一层的结果
		// 也可以用 for 0..queue.len()
		for _, v := range queue {
			level = append(level, v.Val)
			queue = queue[1:]

			if v.Left != nil {
				queue = append(queue, v.Left)
			}

			if v.Right != nil {
				queue = append(queue, v.Right)
			}
		}

		ans = append(ans, level)
	}

	for i, j := 0, len(ans)-1; i < j; i, j = i+1, j-1 {
		ans[i], ans[j] = ans[j], ans[i]
	}

	return ans

}

File: rust/src/tree/107_level_order_bottom.rs


// level_order_bottom 从底部层次遍历，102 的结果反转
pub fn level_order_bottom(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<Vec<i32>> {

    if root.is_none() {
        return vec![];
    }

    let mut ans = vec![];

    use std::collections::VecDeque;
    let mut queue = VecDeque::new();
    queue.push_back(root);


    while !queue.is_empty() {

        let mut level = vec![];

        (0..queue.len()).for_each(|_| {

            let head = queue.pop_front().unwrap().unwrap();
            level.push(head.borrow().val);

            if head.borrow().left.is_some() {
                queue.push_back(head.borrow_mut().left.take());
            }

            if head.borrow().right.is_some() {
                queue.push_back(head.borrow_mut().right.take());
            }
        });

        ans.push(level);

    }

    ans.reverse();
    ans
}